crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -44,38 +47,50 @@ impl TransactionBatch {
     pub fn new(txs: Vec<ProvenTransaction>) -> Result<Self, BuildBatchError> {
         let id = Self::compute_id(&txs);
 
-        // TODO: we need to handle a possibility that a batch contains multiple transactions against
-        //       the same account (e.g., transaction `x` takes account from state `A` to `B` and
-        //       transaction `y` takes account from state `B` to `C`). These will need to be merged
-        //       into a single "update" `A` to `C`.
-        let updated_accounts =
-            txs.iter().map(|tx| (tx.id(), tx.account_update().clone())).collect();
+        let mut updated_accounts = vec![];
+        let mut produced_nullifiers = vec![];
+        let mut unauthenticated_input_notes = BTreeSet::new();
+        for tx in &txs {
+            // TODO: we need to handle a possibility that a batch contains multiple transactions against
+            //       the same account (e.g., transaction `x` takes account from state `A` to `B` and
+            //       transaction `y` takes account from state `B` to `C`). These will need to be merged
+            //       into a single "update" `A` to `C`.
+            updated_accounts.push((tx.id(), tx.account_update().clone()));
+
+            for note in tx.input_notes() {
+                produced_nullifiers.push(note.nullifier());
+                if let Some(note_id) = note.note_id() {
+                    unauthenticated_input_notes.insert(note_id);
+                }
+            }
+        }
 
-        let produced_nullifiers = txs
+        // Populate batch created notes, filtering out notes consumed in the same batch.
+        // Consumed notes are also removed from the unauthenticated input note map.
+        let output_notes: Vec<_> = txs
             .iter()
-            .flat_map(|tx| tx.input_notes().iter())
-            .map(|note| note.nullifier())
+            .flat_map(|tx| tx.output_notes().iter())
+            .filter(|&note| !unauthenticated_input_notes.remove(&note.id()))
+            .cloned()
             .collect();
```

Comment:
@bobbinth:
1. This situation is impossible because such transaction will not be included in the transaction queue at the very beginning of the pipeline, because we require input unauthorized note to be present in in-flight notes or in the store.
2. Good catch, it might be possible.